### PR TITLE
Move angular menu implementation to microservice side

### PR DESCRIPTION
**Pull request by mshima**
Gateways doesn’t need to be regenerated when an entity is added to a microservice.

Related to https://github.com/jhipster/generator-jhipster/issues/17031

---

Please make sure the below checklist is followed for Pull Requests.

- [ ] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [ ] Tests are added where necessary
- [ ] The JDL part is updated if necessary
- [ ] [jhipster-online](https://github.com/jhipster/jhipster-online) is updated if necessary
- [ ] Documentation is added/updated where necessary
- [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (below reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### CODE REVIEW

**Review by mraible — COMMENTED** [review]
I cloned your branch and tried running `jhipster jdl reactive-ms.jdl` with the following JDL:

```
application {
  config {
    baseName gateway
    reactive true
    packageName com.okta.developer.gateway
    applicationType gateway
    authenticationType oauth2
    buildTool gradle
    clientFramework angular
    prodDatabaseType postgresql
    serviceDiscoveryType eureka
    testFrameworks [cypress]
  }
  entities Blog, Post, Tag, Product
}

application {
  config {
    baseName blog
    reactive true
    packageName com.okta.developer.blog
    applicationType microservice
    authenticationType oauth2
    buildTool gradle
    clientFramework angular
    databaseType neo4j
    devDatabaseType neo4j
    prodDatabaseType neo4j
    enableHibernateCache false
    serverPort 8081
    serviceDiscoveryType eureka
  }
  entities Blog, Post, Tag
}

application {
  config {
    baseName store
    reactive true
    packageName com.okta.developer.store
    applicationType microservice
    authenticationType oauth2
    buildTool gradle
    clientFramework angular
    databaseType mongodb
    devDatabaseType mongodb
    prodDatabaseType mongodb
    enableHibernateCache false
    serverPort 8082
    serviceDiscoveryType eureka
  }
  entities Product
}

entity Blog {
  name String required minlength(3)
  handle String required minlength(2)
}

entity Post {
  title String required
  content TextBlob required
  date Instant required
}

entity Tag {
  name String required minlength(2)
}

entity Product {
  title String required
  price BigDecimal required min(0)
  image ImageBlob
}

relationship ManyToOne {
  Blog{user(login)} to User
  Post{blog(name)} to Blog
}

relationship ManyToMany {
  Post{tag(name)} to Tag{post}
}

paginate Post, Tag with infinite-scroll
paginate Product with pagination

microservice Product with store
microservice Blog, Post, Tag with blog

deployment {
  deploymentType docker-compose
  appsFolders [gateway, blog, store]
  dockerRepositoryName "mraible"
}
```

From reading https://github.com/jhipster/generator-jhipster/issues/17031, it seems that if I add `clientFramework` to a microservice, it'll generate microfrontends by default.

I started the gateway to see if it'd work w/o any other apps started. 👍 

> No error on the screen, console tells you what's up:
> ```
> Error loading blog entities
> Error loading store entities
> ```

Then, I started the blog app. It fails to compile:

```
/Users/mraible/Downloads/mf-test/blog/src/main/java/com/okta/developer/blog/config/neo4j/Neo4jMigrations.java:28: error: apply(MigrationContext) in V0001__CreateUsers cannot implement apply(MigrationContext) in Migration
        public void apply(MigrationContext context) throws IOException {
                    ^
  overridden method does not throw IOException
Note: /Users/mraible/Downloads/mf-test/blog/src/main/java/com/okta/developer/blog/config/neo4j/Neo4jMigrations.java uses unchecked or unsafe operations.
```

If I remove `throws IOException` from `Neo4jMigrations` and add try/catches, it compiles but fails to start with:

```
***************************
APPLICATION FAILED TO START
***************************

Description:

The dependencies of some of the beans in the application context form a cycle:

┌─────┐
|  org.springframework.security.config.annotation.web.reactive.ServerHttpSecurityConfiguration
↑     ↓
|  org.springframework.boot.autoconfigure.web.reactive.WebFluxAutoConfiguration$EnableWebFluxConfiguration
└─────┘
```

This is when I realized my app is using Spring Boot 2.6.1. I thought checking out the `main` branch in the `jhipster-bom` project and running `./mvnw install -Dgpg.skip=true` would fix this, but it doesn't. Is there a different process to install the bom locally for Gradle?

**Comment by mraible** [comment]
I ran `npm unlink` and `npm uninstall -g generator-jhipster`, then `npm link` again and that seems to get me the correct version. At least with the gateway:

```
:: JHipster 🤓  :: Running Spring Boot 2.5.7 ::
```

Its `gradle.properties` has this same version of Spring Boot.

```properties
jhipsterDependenciesVersion=7.4.2-SNAPSHOT
# The spring-boot version should match the one managed by
# https://mvnrepository.com/artifact/tech.jhipster/jhipster-dependencies/7.4.2-SNAPSHOT
springBootVersion=2.5.7
```

The same compilation error happens with the blog app and its `build.gradle` matches the gateway's. 

I tried `rm -rf ~/.m2/repository/tech/jhipster` and re-installing the jhipster-bom. It doesn't work.

**Comment by pascalgrimaud** [comment]
@mshima is right. Neo4j is broken since months:
https://github.com/hipster-labs/jhipster-daily-builds/actions/workflows/neo4j.yaml?page=3

cc @atomfrede

**Comment by mraible** [comment]
@michael-simons Do you know how we can fix Neo4jMigrations?

**Comment by michael-simons** [comment]
And there we go…

**Comment by mraible** [comment]
Thanks @michael-simons!

@mshima Even after I fix Neo4J, I'm still getting Spring Boot 2.6.1 pulled in for some reason. If I add:

```yaml
spring:
  main:
    allow-circular-references: true
```

I get a bit further, but it still fails:

```
***************************
APPLICATION FAILED TO START
***************************

Description:

Your project setup is incompatible with our requirements due to the following reasons:

- Spring Boot [2.6.1] is not compatible with this Spring Cloud release train
```

Where does `7.4.2-SNAPSHOT` get pulled in from if it's not the local Maven repo? My local repo has 2.5.7:

```
$ cat ~/.m2/repository/tech/jhipster/jhipster-parent/7.4.2-SNAPSHOT/jhipster-parent-7.4.2-SNAPSHOT.pom| grep spring-boot.version
        <!-- The spring-boot version should be the same as the parent version above -->
        <spring-boot.version>2.5.7</spring-boot.version>
```

**Comment by michael-simons** [comment]
Neo4j-Migrations-Spring-Boot-Starter 1.0.0 uses Boot 2.5, 1.1+ 2.6
Normally Boot's dependency management should make the right choice.
But: You're safe to force-downgrade the Boot dependency in Neo4j-Migrations-Spring-Boot-Starter, there isn't anything 2.6 specific used.

**Comment by mraible** [comment]
It looks like `build.gradle` requires you to make changes to use a snapshot release of jhipster-framework. It's strange because it also looks like if you're using gae, it pulls it in automatically.

```groovy
// import JHipster dependencies BOM
if (!project.hasProperty("gae")) {
    implementation platform("tech.jhipster:jhipster-dependencies:${jhipsterDependenciesVersion}")
}

// Use ", version: jhipsterDependenciesVersion, changing: true" if you want
// to use a SNAPSHOT release instead of a stable release
implementation group: "tech.jhipster", name: "jhipster-framework", {
    exclude group: "org.springframework", module: "spring-webmvc"
}
```

If I modify the suggested block to the following:

```groovy
implementation group: "tech.jhipster", name: "jhipster-framework", version: jhipsterDependenciesVersion, changing: true {
    exclude group: "org.springframework", module: "spring-webmvc"
}
```

It fails too:

```
$ ./gradlew

FAILURE: Build failed with an exception.

* Where:
Build file '/Users/mraible/Downloads/mf-test/blog/build.gradle' line: 192

* What went wrong:
A problem occurred evaluating root project 'blog'.
> No signature of method: java.lang.Boolean.call() is applicable for argument types: (build_dszj7m0zu1av8sl18uiep2jlz$_run_closure15$_closure31) values: [build_dszj7m0zu1av8sl18uiep2jlz$_run_closure15$_closure31@58934f5d]
  Possible solutions: wait(), any(), wait(long), any(groovy.lang.Closure), tap(groovy.lang.Closure), and(java.lang.Boolean)
```

**Comment by michael-simons** [comment]
So, that's surely not something Neo4j-Migrations is causing ;)

Please ping me if there's anything more I can do.
The Boot-Plugin I offer is not much over the core api for the tooling.

**Comment by mshima** [comment]
@mraible happened with springdoc too.
Need to exclude.
https://github.com/jhipster/generator-jhipster/pull/17315#issuecomment-994904044

**Comment by mraible** [comment]
@mshima My local BOM that's installed has the exclusion you mention. It seems like my blog app is not using it. Are you saying we also need to exclude it in the generated app's `build.gradle`?

I tried adding the following, but it doesn't help:

```groovy
implementation ("org.springdoc:springdoc-openapi-webmvc-core") {
    exclude group: "org.springframework.boot"
}
```

I think the main problem is these instructions don't work anymore.

```groovy
// Use ", version: jhipsterDependenciesVersion, changing: true" if you want
// to use a SNAPSHOT release instead of a stable release
implementation group: "tech.jhipster", name: "jhipster-framework", {
    exclude group: "org.springframework", module: "spring-webmvc"
}
```

**Comment by mshima** [comment]
Neo4j + gradle fixed too.

**Comment by mraible** [comment]
@mshima Excluding like you did from the neo4j starter fixes the problem for me locally and I didn't need to modify my `build.gradle` to make it work. Thanks!

**Comment by mraible** [comment]
Continuing my review... 

I started the gateway only and see no entities. If I start the blog and refresh the gateway, Entities > Blog doesn't show up. I kind of expected it too, but I now realize it probably shouldn't show up because the gateway UI is not listening for changes. If I run `npm start` in the gateway, I do see the blog entities.

If I stop the blog, I have a hard time getting the entities to show up again after a restart. The error in the logs is:

```
[WDS] Warnings while compiling.
index.js:135 
{message: '\x1B[0m\x1B[0m\n\x1B[0m\x1B[4m/Users/mraible/Downloads/mf-test/…rnings)\x1B[22m\x1B[39m\x1B[0m\n\x1B[0m\x1B[33m\x1B[1m\x1B[22m\x1B[39m\x1B[0m'}
message: "\u001b[0m\u001b[0m\n\u001b[0m\u001b[4m/Users/mraible/Downloads/mf-test/gateway/src/main/webapp/app/app.module.ts\u001b[24m\u001b[0m\n\u001b[0m  \u001b[2m8:31\u001b[22m  \u001b[33mwarning\u001b
[39m  'SessionStorageService' is defined but never used  \u001b[2m@typescript-eslint/no-unused-vars\u001b[22m\u001b[0m\n\u001b[0m\u001b[0m\n\u001b[0m\u001b[4m/Users/mraible/Downloads/mf-test/gateway/src/main/webapp/app/layouts/navbar/navbar.component.ts\u001b[24m\u001b[0m\n\u001b[0m  \u001b[2m1:49\u001b[22m  \u001b[33mwarning\u001b
[39m  'NgModuleFactory' is defined but never used  \u001b[2m@typescript-eslint/no-unused-vars\u001b[22m\u001b[0m\n\u001b[0m\u001b[0m\n\u001b[0m\u001b[4m/Users/mraible/Downloads/mf-test/gateway/src/main/webapp/app/shared/language/translation.module.ts\u001b[24m\u001b[0m\n\u001b[0m  \u001b[2m4:10\u001b[22m  \u001b[33mwarning\u001b
[39m  'TranslateHttpLoader' is defined but never used  \u001b[2m@typescript-eslint/no-unused-vars\u001b[22m\u001b[0m\n\u001b[0m\u001b[0m\n\u001b[0m\u001b[33m\u001b[1m✖ 3 problems (0 errors, 3 warnings)\u001b[22m\u001b[39m\u001b[0m\n\u001b[0m\u001b[33m\u001b[1m\u001b[22m\u001b[39m\u001b[0m"
[[Prototype]]: Object
```

If I restart the gateway, I can see the entities. Is this supposed to be dynamic and recognize that services are available/unavailable by showing/hiding menu items?

**Comment by mraible** [comment]
For some reason, starting and starting the store works as expected. The menu item is added and removed when I click refresh. For the blog, once I had to shift+reload to get it to show up. The next time, it worked as expected. On the 2nd stop/start with the store, I had to hard reload as well. Probably not a big deal so I'll approve.

File: generators/app/index.js [hunk 1]
```diff
@@ -250,10 +250,15 @@ module.exports = class JHipsterAppGenerator extends BaseBlueprintGenerator {
     });
 
     this.option('microfrontend', {
-      desc: 'Use experimental microfrontend support',
+      desc: 'Force to generate experimental microfrontend support',
```
Comment (mraible) [inline generators/app/index.js:253]:
```suggestion
      desc: 'Force generation of experimental microfrontend support',
```

**Review by mraible — APPROVED** [review]

**Comment by mshima** [comment]
@mraible the microservice needs accessible through `services/(blog/store)`.
So it needs to be up and takes a few seconds (5-10s I think) to be advertised and accessible.
You can watch using admin/gateway to see the advertised services.
Looks pretty stable afterwards.

**Pull request merged**